# Supermarket analysis-Ədalət Sadıqov

## Libraries

In [60]:
import pandas as pd
import numpy as np
import plotly.express as px
from scipy import stats as st
import plotly.graph_objects as go
from plotly.subplots import make_subplots

print('All libraries are ready to use.')


All libraries are ready to use.


## Loading Dataset

In [61]:
data = pd.read_csv('supermarket_sales - Sheet1.csv')

In [62]:
data.head(5)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3


## Exploring data

In [63]:
data.columns = data.columns.str.lower().str.replace(' ', '_')

In [64]:
print(f'Rows: {data.shape[0]}')
print(f'Columns: {data.shape[1]}')
print()
print('Column names')
print(data.columns.to_list())

Rows: 1000
Columns: 17

Column names
['invoice_id', 'branch', 'city', 'customer_type', 'gender', 'product_line', 'unit_price', 'quantity', 'tax_5%', 'total', 'date', 'time', 'payment', 'cogs', 'gross_margin_percentage', 'gross_income', 'rating']


In [65]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   invoice_id               1000 non-null   str    
 1   branch                   1000 non-null   str    
 2   city                     1000 non-null   str    
 3   customer_type            1000 non-null   str    
 4   gender                   1000 non-null   str    
 5   product_line             1000 non-null   str    
 6   unit_price               1000 non-null   float64
 7   quantity                 1000 non-null   int64  
 8   tax_5%                   1000 non-null   float64
 9   total                    1000 non-null   float64
 10  date                     1000 non-null   str    
 11  time                     1000 non-null   str    
 12  payment                  1000 non-null   str    
 13  cogs                     1000 non-null   float64
 14  gross_margin_percentage  1000 non-nu

In [66]:
data.describe()

,unit_price,quantity,tax_5%,total,cogs,gross_margin_percentage,gross_income,rating
count,1000.000000,1000.000000,1000.000000,1000.000000,1000.00000,1000.000000,1000.000000,1000.00000
mean,55.672130,5.510000,15.379369,322.966749,307.58738,4.761905,15.379369,6.97270
std,26.494628,2.923431,11.708825,245.885335,234.17651,0.000000,11.708825,1.71858
min,10.080000,1.000000,0.508500,10.678500,10.17000,4.761905,0.508500,4.00000
25%,32.875000,3.000000,5.924875,124.422375,118.49750,4.761905,5.924875,5.50000
50%,55.230000,5.000000,12.088000,253.848000,241.76000,4.761905,12.088000,7.00000
75%,77.935000,8.000000,22.445250,471.350250,448.90500,4.761905,22.445250,8.50000
max,99.960000,10.000000,49.650000,1042.650000,993.00000,4.761905,49.650000,10.00000


## Data checking

In [67]:
print('Missing Values per columns')
print(data.isnull().sum())
print(f'Total missing values: {data.isnull().sum().sum()}')

Missing Values per columns
invoice_id                 0
branch                     0
city                       0
customer_type              0
gender                     0
product_line               0
unit_price                 0
quantity                   0
tax_5%                     0
total                      0
date                       0
time                       0
payment                    0
cogs                       0
gross_margin_percentage    0
gross_income               0
rating                     0
dtype: int64
Total missing values: 0


In [68]:
duplicate = data.duplicated().sum()
print(f'Duplicated rows: {duplicate}')

Duplicated rows: 0


In [69]:
#invoice id should be unique 1 id per transaction

print(f'Unique invoice_id`s: {data['invoice_id'].nunique()}/{len(data)}')

Unique invoice_id`s: 1000/1000


In [70]:
data.columns

Index(['invoice_id', 'branch', 'city', 'customer_type', 'gender',
       'product_line', 'unit_price', 'quantity', 'tax_5%', 'total', 'date',
       'time', 'payment', 'cogs', 'gross_margin_percentage', 'gross_income',
       'rating'],
      dtype='str')

No missing values and no duplicate rows — the dataset is clean. No imputation
or row-dropping is required before moving forward.

In [71]:
data.describe(include='all')

,invoice_id,branch,city,customer_type,gender,product_line,unit_price,quantity,tax_5%,total,date,time,payment,cogs,gross_margin_percentage,gross_income,rating
count,1000,1000,1000,1000,1000,1000,1000.000000,1000.000000,1000.000000,1000.000000,1000,1000,1000,1000.00000,1000.000000,1000.000000,1000.00000
unique,1000,3,3,2,2,6,NaN,NaN,NaN,NaN,89,506,3,NaN,NaN,NaN,NaN
top,750-67-8428,A,Yangon,Member,Female,Fashion accessories,NaN,NaN,NaN,NaN,2/7/2019,19:48,Ewallet,NaN,NaN,NaN,NaN
freq,1,340,340,501,501,178,NaN,NaN,NaN,NaN,20,7,345,NaN,NaN,NaN,NaN
mean,NaN,NaN,NaN,NaN,NaN,NaN,55.672130,5.510000,15.379369,322.966749,NaN,NaN,NaN,307.58738,4.761905,15.379369,6.97270
std,NaN,NaN,NaN,NaN,NaN,NaN,26.494628,2.923431,11.708825,245.885335,NaN,NaN,NaN,234.17651,0.000000,11.708825,1.71858
min,NaN,NaN,NaN,NaN,NaN,NaN,10.080000,1.000000,0.508500,10.678500,NaN,NaN,NaN,10.17000,4.761905,0.508500,4.00000
25%,NaN,NaN,NaN,NaN,NaN,NaN,32.875000,3.000000,5.924875,124.422375,NaN,NaN,NaN,118.49750,4.761905,5.924875,5.50000
50%,NaN,NaN,NaN,NaN,NaN,NaN,55.230000,5.000000,12.088000,253.848000,NaN,NaN,NaN,241.76000,4.761905,12.088000,7.00000
75%,NaN,NaN,NaN,NaN,NaN,NaN,77.935000,8.000000,22.445250,471.350250,NaN,NaN,NaN,448.90500,4.761905,22.445250,8.50000


## Data cleaning

In [72]:
data_copy = data.copy()

In [73]:
data_copy['date'] = pd.to_datetime(data_copy['date'],format="%m/%d/%Y")

Convert Date to a real datetime column

In [74]:
data_copy["hour"] = pd.to_datetime(data_copy["time"], format="%H:%M").dt.hour

Extract the hour of day from Time

In [75]:
data_copy['month'] = data_copy['date'].dt.month
data_copy['dayofweek'] = data_copy['date'].dt.day_name().str.lower()
data_copy['weekend'] = data_copy['dayofweek'].isin(['saturday','sunday'])

In [76]:
data_copy

,invoice_id,branch,city,customer_type,gender,product_line,unit_price,quantity,tax_5%,total,...,time,payment,cogs,gross_margin_percentage,gross_income,rating,hour,month,dayofweek,weekend
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,...,13:08,Ewallet,522.83,4.761905,26.1415,9.1,13,1,saturday,True
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,...,10:29,Cash,76.40,4.761905,3.8200,9.6,10,3,friday,False
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,...,13:23,Credit card,324.31,4.761905,16.2155,7.4,13,3,sunday,True
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,...,20:33,Ewallet,465.76,4.761905,23.2880,8.4,20,1,sunday,True
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,...,10:37,Ewallet,604.17,4.761905,30.2085,5.3,10,2,friday,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,233-67-5758,C,Naypyitaw,Normal,Male,Health and beauty,40.35,1,2.0175,42.3675,...,13:46,Ewallet,40.35,4.761905,2.0175,6.2,13,1,tuesday,False
996,303-96-2227,B,Mandalay,Normal,Female,Home and lifestyle,97.38,10,48.6900,1022.4900,...,17:16,Ewallet,973.80,4.761905,48.6900,4.4,17,3,saturday,True
997,727-02-1313,A,Yangon,Member,Male,Food and beverages,31.84,1,1.5920,33.4320,...,13:22,Cash,31.84,4.761905,1.5920,7.7,13,2,saturday,True
998,347-56-2442,A,Yangon,Normal,Male,Home and lifestyle,65.82,1,3.2910,69.1110,...,15:33,Cash,65.82,4.761905,3.2910,4.1,15,2,friday,False


In [77]:
print("New columns added: Hour, Month, DayOfWeek, Weekend")
data_copy[["date", "time", "hour", "month", "dayofweek", "weekend"]].head()

New columns added: Hour, Month, DayOfWeek, Weekend


,date,time,hour,month,dayofweek,weekend
0,2019-01-05,13:08,13,1,saturday,True
1,2019-03-08,10:29,10,3,friday,False
2,2019-03-03,13:23,13,3,sunday,True
3,2019-01-27,20:33,20,1,sunday,True
4,2019-02-08,10:37,10,2,friday,False


`date`and`time` are currently as text. so we convert them to proper date/time types

## Analysis

### Group analysis

In [78]:
data_copy.columns.to_list()

['invoice_id',
 'branch',
 'city',
 'customer_type',
 'gender',
 'product_line',
 'unit_price',
 'quantity',
 'tax_5%',
 'total',
 'date',
 'time',
 'payment',
 'cogs',
 'gross_margin_percentage',
 'gross_income',
 'rating',
 'hour',
 'month',
 'dayofweek',
 'weekend']

In [79]:
branch = data_copy.groupby('branch').agg(
    total_sales = ('total','sum'),
    order_count = ('invoice_id','count'),
    avg_order_value = ('total','mean')
).round(2).reset_index()

branch

,branch,total_sales,order_count,avg_order_value
0,A,106200.37,340,312.35
1,B,106197.67,332,319.87
2,C,110568.71,328,337.10


In [80]:
branch = branch.sort_values('total_sales',ascending= False)

In [81]:
print('Sales summary by branch')
branch

Sales summary by branch


,branch,total_sales,order_count,avg_order_value
2,C,110568.71,328,337.10
0,A,106200.37,340,312.35
1,B,106197.67,332,319.87


In [82]:
product = data_copy.groupby('product_line').agg(
    total_sales = ('total','sum'),
    order_count = ('invoice_id','count'),
    avg_order_value = ('total','mean')
).round(2).reset_index()

product

,product_line,total_sales,order_count,avg_order_value
0,Electronic accessories,54337.53,170,319.63
1,Fashion accessories,54305.90,178,305.09
2,Food and beverages,56144.84,174,322.67
3,Health and beauty,49193.74,152,323.64
4,Home and lifestyle,53861.91,160,336.64
5,Sports and travel,55122.83,166,332.07


In [83]:
product = product.sort_values('total_sales',ascending= False)

In [84]:
print('Sales summary by Product line:')
product

Sales summary by Product line:


,product_line,total_sales,order_count,avg_order_value
2,Food and beverages,56144.84,174,322.67
5,Sports and travel,55122.83,166,332.07
0,Electronic accessories,54337.53,170,319.63
1,Fashion accessories,54305.90,178,305.09
4,Home and lifestyle,53861.91,160,336.64
3,Health and beauty,49193.74,152,323.64


In [85]:
customer = data_copy.groupby("customer_type").agg(
    total_sales=("total", "sum"),
    order_count=("invoice_id", "count"),
    avg_order_value=("total", "mean"),
).round(2).reset_index()

customer

,customer_type,total_sales,order_count,avg_order_value
0,Member,164223.44,501,327.79
1,Normal,158743.30,499,318.12


In [86]:
print('Sales summary bt cutomer type')
customer

Sales summary bt cutomer type


,customer_type,total_sales,order_count,avg_order_value
0,Member,164223.44,501,327.79
1,Normal,158743.30,499,318.12


In [87]:
gender = data_copy.groupby("gender").agg(
    total_sales=("total", "sum"),
    order_count=("invoice_id", "count"),
    avg_order_value=("total", "mean"),
).round(2).reset_index()

gender

,gender,total_sales,order_count,avg_order_value
0,Female,167882.92,501,335.10
1,Male,155083.82,499,310.79


In [88]:
print("Sales summary by Gender:")
gender

Sales summary by Gender:


,gender,total_sales,order_count,avg_order_value
0,Female,167882.92,501,335.10
1,Male,155083.82,499,310.79


Order counts are almost identical across the three branches, so any
difference in total sales comes from average order value, not from order count. Whether this


### Hourly analysis

In [89]:
hourly_sales = data_copy.groupby('hour').agg(
    total_sales = ('total','sum'),
    order_count = ('invoice_id','count')
).round(2).reset_index()

hourly_sales

,hour,total_sales,order_count
0,10,31421.48,101
1,11,30377.33,90
2,12,26065.88,89
3,13,34723.23,103
4,14,30828.40,83
5,15,31179.51,102
6,16,25226.32,77
7,17,24445.22,74
8,18,26030.34,93
9,19,39699.51,113


In [90]:
hourly_sales = hourly_sales.sort_values('total_sales',ascending= False)

In [91]:
print('Top 5 busiest hours by total sales:')

print(hourly_sales.head(5).to_string(index= False))

Top 5 busiest hours by total sales:
 hour  total_sales  order_count
   19     39699.51          113
   13     34723.23          103
   10     31421.48          101
   15     31179.51          102
   14     30828.40           83


In [92]:
busiest_hour = hourly_sales.iloc[0]['hour']
print(f'\nBusiest hour: {int(busiest_hour)}:00')


Busiest hour: 19:00


Sales peak at 19:00 (evening shopping), with a secondary peak at 13:00 (lunch
break).

### Product lines by revenue

In [93]:
top_product = product.iloc[0]
bot_product = product.iloc[-1]


print('----Top product----')
display(top_product)
print('----Bottom product----')
display(bot_product)

----Top product----


product_line       Food and beverages
total_sales                  56144.84
order_count                       174
avg_order_value                322.67
Name: 2, dtype: object

----Bottom product----


product_line       Health and beauty
total_sales                 49193.74
order_count                      152
avg_order_value               323.64
Name: 3, dtype: object

In [94]:
print(f'Highest revenue product line: {top_product['product_line']}'
      f'({top_product['total_sales']:.2f})$')

print(f'Lowest revenue product line: {bot_product['product_line']}'
      f'({bot_product['total_sales']:.2f})$')

Highest revenue product line: Food and beverages(56144.84)$
Lowest revenue product line: Health and beauty(49193.74)$


In [95]:
gap = top_product["total_sales"] - bot_product["total_sales"]
gap_pct = (gap / bot_product["total_sales"]) * 100
print(f"\nGap: ${gap:,.2f} ({gap_pct:.1f}%)")


Gap: $6,951.10 (14.1%)


The gap is small (~14%) across 6 product lines — sales volume is not
sharply concentrated in one category, giving us a reason to say this business has a balanced product portfolio.

## Visuals

In [96]:
product_sorted = product.sort_values("total_sales", ascending=True)

fig_product = px.bar(
    product_sorted,
    x="total_sales",
    y="product_line",
    orientation="h",
    text="total_sales",
    color="total_sales",
    color_continuous_scale="Blues",
)

fig_product.update_traces(
    textposition="outside"
)

fig_product.update_layout(
    title="Total Sales by Product Line",
    xaxis_title="Total Sales ($)",
    yaxis_title="Product Line",
    coloraxis_showscale=False,
    template="plotly_white",
    height=500,
)

fig_product.show()

Food and Beverages is the top-revenue category, but the gap to the lowest
(Health and Beauty) is only ~14% — no single category dominates the sales.

In [97]:
customer_counts = data_copy['customer_type'].value_counts()

fig_customer = px.pie(
    names=customer_counts.index,
    values=customer_counts.values,
    hole=0.45,
    color_discrete_sequence=px.colors.qualitative.Set2,
    title="Customer Type Distribution",
)

fig_customer.update_traces(
    textinfo="label+percent",
    textposition="inside",
    hovertemplate="<b>%{label}</b><br>Orders: %{value}<br>Share: %{percent}<extra></extra>",
    marker=dict(line=dict(color="white", width=2)),
)

fig_customer.update_layout(
    template="plotly_white",
    title_x=0.5,
    legend_title_text="Customer Type",
    height=500,
)


fig_customer.show()

Member and Normal types are diveded almost exactly (50.1% / 49.9% ). Means Half of the customer base has not joind to loyalty program

In [98]:
hourly_sales_sorted = hourly_sales.sort_values("hour")

fig_hour = px.line(
    hourly_sales_sorted,
    x="hour",
    y="total_sales",
    markers=True,
    title="Total Sales by Hour of Day",
    labels={
        "hour": "Hour of Day",
        "total_sales": "Total Sales ($)"
    }
)

fig_hour.update_layout(
    template="plotly_white",
    xaxis=dict(
        tickmode="linear",
        dtick=1
    )
)

fig_hour.show()

Sales ramp up from the morning, show two clear peaks (13:00 and 19:00), and dip
in the mid-afternoon (15:00–17:00). Staff scheduling should track these two demand windows.

In [99]:
sales_heatmap = data_copy.pivot_table(
    index="branch",
    columns="product_line",
    values="total",
    aggfunc="sum"
)

fig_heatmap = px.imshow(
    sales_heatmap,
    text_auto=".2f",
    color_continuous_scale="Blues",
    aspect="auto",
    labels={
        "x": "Product Line",
        "y": "Branch",
        "color": "Total Sales ($)"
    },
    title="Total Sales by Branch and Product Line"
)

fig_heatmap.update_layout(
    template="plotly_white",
    height=500
)

fig_heatmap.show()

From the heatmap there is no dominant  branch all of branches.
- Branch C looks stronger in `Food and beverages`
- Branch A looks stornger in `Home and lifestyle`

## Gross margin analysis

In [100]:
unique_margins = data_copy['gross_margin_percentage'].unique()

print(f'Unique values in gross margin percentage: {unique_margins}')

Unique values in gross margin percentage: [4.76190476]


In [101]:
print(f'Standard deviation : {data_copy['gross_margin_percentage'].std()}')

Standard deviation : 0.0


`gross margin percentage` is identical (4.76%) on all 1,000 rows,
because in this dataset tax is always exactly 5% of `cogs`, so `Total = cogs × 1.05`.
That ratio can never vary mathematically, so it carries no differentiating information by product
line — grouping on it as-is would be misleading.

In [102]:
margin_summary = data_copy.groupby('product_line').agg(
    total_gross_income = ('gross_income','sum'),
    total_sales = ('total','sum')
).round(2).reset_index()


margin_summary

,product_line,total_gross_income,total_sales
0,Electronic accessories,2587.50,54337.53
1,Fashion accessories,2586.00,54305.90
2,Food and beverages,2673.56,56144.84
3,Health and beauty,2342.56,49193.74
4,Home and lifestyle,2564.85,53861.91
5,Sports and travel,2624.90,55122.83


In [103]:
margin_summary['margin_pct_contast'] = data_copy['gross_margin_percentage'].iloc[0]

mragin_summary = margin_summary.sort_values('total_gross_income',ascending= False)
print("Gross income summary by product line:")
margin_summary

Gross income summary by product line:


,product_line,total_gross_income,total_sales,margin_pct_contast
0,Electronic accessories,2587.50,54337.53,4.761905
1,Fashion accessories,2586.00,54305.90,4.761905
2,Food and beverages,2673.56,56144.84,4.761905
3,Health and beauty,2342.56,49193.74,4.761905
4,Home and lifestyle,2564.85,53861.91,4.761905
5,Sports and travel,2624.90,55122.83,4.761905


`total_gross_income` ranks in exactly the same order as `total_sales` — expected,
since the margin is fixed. Within this dataset, profitability differences are driven entirely by
sales volume, 

## Statistical analysis

In [104]:
def interperet(p_value:float,alpha = 0.05):
    if p_value < alpha:
        return f"SIGNIFICANT (p={p_value:.4f})"
    else:
        return f"NOT significant (p={p_value:.4f})"

### Correlation analysis

In [105]:
data_copy.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 21 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   invoice_id               1000 non-null   str           
 1   branch                   1000 non-null   str           
 2   city                     1000 non-null   str           
 3   customer_type            1000 non-null   str           
 4   gender                   1000 non-null   str           
 5   product_line             1000 non-null   str           
 6   unit_price               1000 non-null   float64       
 7   quantity                 1000 non-null   int64         
 8   tax_5%                   1000 non-null   float64       
 9   total                    1000 non-null   float64       
 10  date                     1000 non-null   datetime64[us]
 11  time                     1000 non-null   str           
 12  payment                  1000 non-null   str  

In [106]:
numeric_cols = ['unit_price','quantity','total','cogs','gross_income','rating']

corr_matrix = data_copy[numeric_cols].corr().round(2)

print('Correlation matrix')
corr_matrix

Correlation matrix


,unit_price,quantity,total,cogs,gross_income,rating
unit_price,1.00,0.01,0.63,0.63,0.63,-0.01
quantity,0.01,1.00,0.71,0.71,0.71,-0.02
total,0.63,0.71,1.00,1.00,1.00,-0.04
cogs,0.63,0.71,1.00,1.00,1.00,-0.04
gross_income,0.63,0.71,1.00,1.00,1.00,-0.04
rating,-0.01,-0.02,-0.04,-0.04,-0.04,1.00


In [107]:
fig_corr = px.imshow(
    corr_matrix,
    text_auto=".2f",
    color_continuous_scale="RdBu_r",
    zmin=-1,
    zmax=1,
    title="Correlation Heatmap"
)

fig_corr.update_layout(
    template="plotly_white",
    height=600,
    xaxis_title="Variables",
    yaxis_title="Variables"
)

fig_corr.show()

In [108]:
r,p = st.pearsonr(data_copy['total'],data_copy['rating'])

print(f'Pearson correlation, total vs rating: r= {r:.4f},{interperet(p)}')

Pearson correlation, total vs rating: r= -0.0364,NOT significant (p=0.2496)


`Total`, `cogs`, and `gross income` are perfectly correlated (r=1.0) due to the fixed 5% margin. More importantly, `Rating` is essentially uncorrelated with spending `(r≈-0.04)`, suggesting customer satisfaction is unrelated to spending or purchase volume. Correlation indicates association, not causation.

### Shapiro-Wilk

In [109]:
for col in ['total','rating']:
    stat,p = st.shapiro(data_copy[col])
    print(f'{col}: W = {stat:.4f}, {interperet(p)}')

total: W = 0.9088, SIGNIFICANT (p=0.0000)
rating: W = 0.9582, SIGNIFICANT (p=0.0000)


`Total` and `Rating` are not normally distributed `(p < 0.001)`. Therefore, non-parametric tests are used as the main results. With n=1,000, parametric tests give similar results.

### Spending: Member vs normal customers(t-test/Mann-Whitney U)

In [110]:
member_total = data_copy.loc[data_copy['customer_type'] == 'Member', 'total']
normal_total = data_copy.loc[data_copy['customer_type'] == 'Normal', 'total']

In [111]:
t_stat, t_p = st.ttest_ind(member_total,normal_total)
u_stat, u_p = st.mannwhitneyu(member_total,normal_total)

In [112]:
print(f"T-test: t={t_stat:.4f}, {interperet(t_p)}")
print(f"Mann-Whitney: U={u_stat:.1f}, {interperet(u_p)}")

print(data_copy.groupby("customer_type")["total"].agg(["mean", "median", "std"]).round(2))

T-test: t=0.6215, NOT significant (p=0.5344)
Mann-Whitney: U=127453.0, NOT significant (p=0.5912)
                 mean  median     std
customer_type                        
Member         327.79  266.03  247.99
Normal         318.12  237.43  243.90


Members spend slightly more ($327.79 vs $318.12), but the difference is not statistically significant (p > 0.5). So, membership does not lead to higher spending per visit.

### spending: Weekend vs Weekday

In [113]:
data_copy

,invoice_id,branch,city,customer_type,gender,product_line,unit_price,quantity,tax_5%,total,...,time,payment,cogs,gross_margin_percentage,gross_income,rating,hour,month,dayofweek,weekend
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,...,13:08,Ewallet,522.83,4.761905,26.1415,9.1,13,1,saturday,True
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,...,10:29,Cash,76.40,4.761905,3.8200,9.6,10,3,friday,False
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,...,13:23,Credit card,324.31,4.761905,16.2155,7.4,13,3,sunday,True
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,...,20:33,Ewallet,465.76,4.761905,23.2880,8.4,20,1,sunday,True
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,...,10:37,Ewallet,604.17,4.761905,30.2085,5.3,10,2,friday,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,233-67-5758,C,Naypyitaw,Normal,Male,Health and beauty,40.35,1,2.0175,42.3675,...,13:46,Ewallet,40.35,4.761905,2.0175,6.2,13,1,tuesday,False
996,303-96-2227,B,Mandalay,Normal,Female,Home and lifestyle,97.38,10,48.6900,1022.4900,...,17:16,Ewallet,973.80,4.761905,48.6900,4.4,17,3,saturday,True
997,727-02-1313,A,Yangon,Member,Male,Food and beverages,31.84,1,1.5920,33.4320,...,13:22,Cash,31.84,4.761905,1.5920,7.7,13,2,saturday,True
998,347-56-2442,A,Yangon,Normal,Male,Home and lifestyle,65.82,1,3.2910,69.1110,...,15:33,Cash,65.82,4.761905,3.2910,4.1,15,2,friday,False


In [114]:
weekend_total = data_copy.loc[data_copy['weekend'], 'total']
weekday_total = data_copy.loc[~data_copy['weekend'], 'total']

In [115]:
weekend_total

0       548.9715
2       340.5255
3       489.0480
7       772.3800
11      107.1420
         ...    
978      26.7225
980     250.2780
992     121.8630
996    1022.4900
997      33.4320
Name: total, Length: 297, dtype: float64

In [116]:
weekday_total

1       80.2200
4      634.3785
5      627.6165
6      433.6920
8       76.1460
         ...   
993    183.6450
994     63.9975
995     42.3675
998     69.1110
999    649.2990
Name: total, Length: 703, dtype: float64

In [117]:
t_stat, t_p = st.ttest_ind(weekend_total, weekday_total)
u_stat, u_p = st.mannwhitneyu(weekend_total, weekday_total)

In [119]:
print(f"Independent t-test:  t={t_stat:.4f}, {interperet(t_p)}")
print(f"Mann-Whitney U:      U={u_stat:.1f}, {interperet(u_p)}")

Independent t-test:  t=1.3114, NOT significant (p=0.1900)
Mann-Whitney U:      U=111091.5, NOT significant (p=0.1086)


In [120]:
print(data_copy.groupby("weekend")["total"].agg(["mean", "median", "std", "count"]).round(2))

           mean  median     std  count
weekend                               
False    316.34  235.68  246.12    703
True     338.65  295.69  245.02    297


**Finding:** Weekend average order value ($338.65) is higher than weekday ($316.34), but
again **not statistically significant** (p≈0.19 / p≈0.11). With n=1,000 spread over three months.

### Chi-Square Tests 

In [121]:
def chi_square_test(col1, col2):
    table = pd.crosstab(data_copy[col1], data_copy[col2])
    chi2, p, dof, _ = st.chi2_contingency(table)
    print(f"{col1} x {col2}: chi2={chi2:.2f}, p={p:.4f}")

In [123]:
chi_square_test("branch", "product_line")
chi_square_test("gender", "product_line")
chi_square_test("customer_type", "gender")

branch x product_line: chi2=11.56, p=0.3156
gender x product_line: chi2=5.74, p=0.3319
customer_type x gender: chi2=1.44, p=0.2295


Finding: None of the three pairs show a significant association (p > 0.05):

- Branch × Product line: no meaningful branch-based preference.
- Gender × Product line: no significant gender-based preference.
- Customer type × Gender: membership is independent of gender.

The heatmap patterns are likely random variation rather than real relationships.

### Outlier detection

In [126]:
q1 = data_copy["total"].quantile(0.25)
q3 = data_copy["total"].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - 1.5 * iqr
upper_bound = q3 + 1.5 * iqr


In [127]:

outliers = data_copy[(data_copy["total"] < lower_bound) | (data_copy["total"] > upper_bound)]

print(f"Q1={q1:.2f}, Q3={q3:.2f}, IQR={iqr:.2f}")
print(f"Lower bound={lower_bound:.2f}, Upper bound={upper_bound:.2f}")
print(f"Outlier count: {len(outliers)} ({len(outliers) / len(data_copy) * 100:.2f}% of rows)")


Q1=124.42, Q3=471.35, IQR=346.93
Lower bound=-395.97, Upper bound=991.74
Outlier count: 9 (0.90% of rows)


Only 9 rows (0.90%) fall outside the IQR bounds on `Total`. This is a very low
outlier rate — no special outlier-handling (capping, removal) is needed before further analysis.

## Additional analsis: payment method

In [128]:
data_copy.columns.to_list()

['invoice_id',
 'branch',
 'city',
 'customer_type',
 'gender',
 'product_line',
 'unit_price',
 'quantity',
 'tax_5%',
 'total',
 'date',
 'time',
 'payment',
 'cogs',
 'gross_margin_percentage',
 'gross_income',
 'rating',
 'hour',
 'month',
 'dayofweek',
 'weekend']

In [130]:
payment_method = data_copy.groupby('payment').agg(
    total_sales = ('total','sum'),
    avg_order_value = ('total', 'mean'),
    order_count = ('invoice_id','count')
).round(2).sort_values('total_sales', ascending= False).reset_index()

payment_method

,payment,total_sales,avg_order_value,order_count
0,Cash,112206.57,326.18,344
1,Ewallet,109993.11,318.82,345
2,Credit card,100767.07,324.01,311


In [132]:
print("Sales summary by Payment method:")
print(payment_method)

chi_square_test("payment", "branch")

Sales summary by Payment method:
       payment  total_sales  avg_order_value  order_count
0         Cash    112206.57           326.18          344
1      Ewallet    109993.11           318.82          345
2  Credit card    100767.07           324.01          311
payment x branch: chi2=3.30, p=0.5090


Usage is nearly evenly split across Cash, Ewallet, and Credit card (roughly
310–345 orders each), and the chi-square test confirms payment method is independent of branch
(p≈0.51) — no branch relies disproportionately on one payment type.

## Final Conclusion

Supermarket Sales Analysis 

1. Overall picture

Total sales are 322,966.75 and average order value (AOV) is 322.97.
None of the differences across branch, product line, customer type, payment method, gender or weekend/weekday is statistically significant.
The visible gaps are within random variation:
Branch C: +4.1% (p=0.81).
Food & beverages vs Health & beauty: 14.1% gap (p=0.74). Health & beauty trails because of fewer orders, not lower AOV.
Female AOV: +7.8% (p=0.10).
Member AOV: +3.0% (p=0.53).
Weekend AOV: +7.1% (p=0.19 / 0.11).

2. The only signal worth following up

19:00 has 113 orders and 39,700 in sales (12.3% of total).
The unadjusted p=0.018 becomes ≈0.20 after correcting for 11 hourly comparisons. It is a hypothesis, not a finding.